pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/ast/ApexTreeBuilder.kt

```diff
@@ -0,0 +1,137 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.ast
+
+import net.sourceforge.pmd.annotation.InternalApi
+import net.sourceforge.pmd.lang.apex.ApexParserOptions
+import net.sourceforge.pmd.lang.ast.ParseException
+import net.sourceforge.pmd.lang.ast.SourceCodePositioner
+
+import com.google.summit.ast.CompilationUnit
+import com.google.summit.ast.Identifier
+import com.google.summit.ast.Node
+import com.google.summit.ast.SourceLocation
+import com.google.summit.ast.TypeRef
+import com.google.summit.ast.declaration.ClassDeclaration
+import com.google.summit.ast.declaration.EnumDeclaration
+import com.google.summit.ast.declaration.InterfaceDeclaration
+import com.google.summit.ast.declaration.TriggerDeclaration
+import com.google.summit.ast.declaration.TypeDeclaration
+import com.google.summit.ast.modifier.KeywordModifier
+import com.google.summit.ast.modifier.Modifier
+
+@Deprecated("internal")
+@InternalApi
+@Suppress("DEPRECATION")
+class ApexTreeBuilder(val sourceCode: String, val parserOptions: ApexParserOptions) {
+    private val sourceCodePositioner = SourceCodePositioner(sourceCode)
+
+    /** Builds and returns an [ApexNode] AST corresponding to the given [root] node. */
+    fun buildTree(root: CompilationUnit): ApexRootNode<TypeDeclaration> {
+        // Build tree
+        val result =
+            build(root, parent = null) as? ApexRootNode<TypeDeclaration>
+                ?: throw ParseException("Unable to build tree")
+
+        // Call additional methods
+        callAdditional(result)
+
+        return result
+    }
+
+    /** Calls additional methods for each node in [root] using a post-order traversal. */
+    private fun callAdditional(root: ApexNode<*>) =
+        root.jjtAccept(
+            object : ApexParserVisitorAdapter() {
+                override fun visit(node: ApexNode<*>?, data: Any?): Any? =
+                    super.visit(node, data).also {
+                        if (node is AbstractApexNode<*>) {
+                            node.handleSourceCode(sourceCode)
+                            node.calculateLineNumbers(sourceCodePositioner)
+                        }
+                    }
+            },
+            null
+        )
+
+    /**
+     * Builds an [ApexNode] wrapper for [node].
+     *
+     * Sets the parent of the resulting [ApexNode] to [parent], if it's not `null`.
+     */
+    private fun build(node: Node?, parent: ApexNode<*>?): AbstractApexNode<*>? {
+        val wrapper: AbstractApexNode<*>? =
```

Comment:
I think this should return ApexNode instead of AbstractApexNode.
From what I understand AbstractApexNode is just an abstract class that sits between the ApexNode interface and actual implementations. Imho, client code should refer to apex nodes using the interface type.

Reply:
`AbstractApexNode` was a remnant from the `wrapper?.handleSourceCode` and `wrapper?.calculateLineNumbers` calls that I recently removed from this function, so it can be replaced with `ApexNode` now. Thanks for pointing this out.